### ISSUE TITLE
[Bug]: DDP is going blblblblblblblbl

### ISSUE THREAD

**Issue by TParcollet** [speechbrain/speechbrain#1802]
### Describe the bug

On certain recipe, and with certain models, DDP with X GPUs is slower than mono GPU

### Expected behaviour

Should be (X - loss factor) times faster with X the number of GPUs.

### To Reproduce

Simply run the CommonVoice/ASR/CTC/train_with_wav2vec.py with and without DDP. The expected training time will be longer with DDP.

HOWEVER, if we add self.no_sync() to the fit_batch around compute_forward(): then it works as intended. @RuABraun worked a bit with this no_sync() stuff, maybe he could help? I don't understand why models would be syncing during the forward though ...

**Comment by AsuMagic** [comment]
Adel and I investigated the issue a little and it seems that the problem boils down to the fact that we also need to include the **forward** pass (not just the backward pass) in the `no_sync` context manager.

So only at the last forward/backward should the gradient be synchronized. If you omit the `no_sync` context manager for the **forward**, PyTorch seems to call into some code, `prepare_for_backward`, which presumably induces a bunch of synchronization (either immediately or later).

The PyTorch documentation says so and a [HuggingFace Accelerate tutorial](https://huggingface.co/docs/accelerate/concept_guides/gradient_synchronization) mention the same thing. The performance degradation HF notes is similar to what you are observing (i.e. worse than single-GPU).

So what we are doing in SB right now is incorrect. It shouldn't cause issues other than unnecessary slowdowns, though. Considering the scope of #2010 which might be very related, maybe it should be implemented there? @Adel-Moumen

**Comment by RuABraun** [comment]
Interesting catch! Didn't realize this and the torch docs didn't have that warning at the time.

**Comment by Adel-Moumen** [comment]
Fixed in SB 1.0 :p (see changes in speechbrain.core.py and more precisely the fit_batch function).
